README.md
[![Actions Status](https://github.com/skaji/Distribution-Metadata/actions/workflows/test.yml/badge.svg)](https://github.com/skaji/Distribution-Metadata/actions)

# NAME

Distribution::Metadata - gather distribution metadata in local

# SYNOPSIS

    use Distribution::Metadata;

    my $info = Distribution::Metadata->new_from_module("LWP::UserAgent");

    print $info->name;      # libwww-perl
    print $info->version;   # 6.13
    print $info->distvname; # libwww-perl-6.13
    print $info->author;    # ETHER
    print $info->pathname;  # E/ET/ETHER/libwww-perl-6.13.tar.gz

    print $info->main_module;         # LWP
    print $info->main_module_version; # 6.13
    print $info->main_module_file;    # path of LWP.pm

    print $info->packlist;       # path of .packlist
    print $info->meta_directory; # path of .meta directory
    print $info->install_json;   # path of install.json
    print $info->mymeta_json;    # path of MYMETA.json

    my $files = $info->files; # files which are listed in .packlist

    my $install_json_hash = $info->install_json_hash;
    my $mymeta_json_hash  = $info->mymeta_json_hash;

# DESCRIPTION

(**CAUTION**: This module is still in development phase. API will change without notice.)

Sometimes we want to know:
_Where this module comes from? Which distribution does this module belong to?_

Since [cpanm](https://metacpan.org/pod/cpanm) 1.5000 (released 2011.10.13),
it installs not only modules but also their meta data.
So we can answer that questions!

Distribution::Metadata gathers distribution metadata in local.
That is, this module tries to gather

- main module name, version, file
- `.packlist` file
- `.meta` directory
- `install.json` file
- `MYMETA.json` file

Please note that as mentioned above, **this module deeply depends on cpanm behavior**.
If you install cpan modules by hands or some cpan clients other than cpanm,
this module won't work.

# HOW IT WORKS

Let me explain how `$class->new_from_module($module, inc => $inc)` works.

- Get `$module_file` by

        Module::Metadata->new_from_module($module, inc => $inc)->filename.

- Find `$packlist` in which `$module_file` is listed.
- From `$packlist` pathname (eg: ...auto/LWP/.packlist), determine `$main_module` and main module search directory `$lib`.
- Get `$main_module_version` by

        Module::Metadata->new_from_module($main_module, inc => [$lib, "$lib/$Config{archname}"])->version

- Find install.json that has "name" eq `$main_module`, and provides `$main_module` with version `$main_module_version`.
- Get .meta directory and MYMETA.json with install.json.

## CONSTRUCTORS

- `my $info = $class->new_from_module($module, inc => \@dirs, fill_archlib => $bool)`

    Create Distribution::Metadata instance from module name.

    You can append `inc` argument
    to specify module/packlist/meta search paths. Default is `\@INC`.

    Also you can append `fill_archlib` argument
    so that archlibs are automatically added to `inc` if missing.

    Please note that, even if the module cannot be found,
    `new_from_module` returns a Distribution::Metadata instance.
    However almost all methods returns false for such objects.
    If you want to know whether the distribution was found or not, try:

        my $info = $class->new_from_module($module);

        if ($info->packlist) {
            # found
        } else {
            # not found
        }

- `my $info = $class->new_from_file($file, inc => \@dirs, fill_archlib => $bool)`

    Create Distribution::Metadata instance from file path.
    You can append `inc` and `fill_archlib` arguments too.

    Also `new_from_file` retunes a Distribution::Metadata instance,
    even if file cannot be found.

## METHODS

Please note that the following methods return false
when appropriate modules or files cannot be found.

- `my $name = $info->name (alias: $info->dist)`

    distribution name (eg: `libwww-perl`)

- `my $version = $info->version`

    distribution version (eg: `6.13`)

- `my $distvname = $info->distvname`

    distribution vname (eg: `libwww-perl-6.13`)

- `my $author = $info->author (alias: $info->cpanid)`

    distribution author (eg: `ETHER`)

- `my $pathname = $info->pathname`

    distribution pathname (eg: `E/ET/ETHER/libwww-perl-6.13.tar.gz`)

- `my $file = $info->packlist`

    `.packlist` file path

- `my $dir = $info->meta_directory`

    `.meta` directory path

- `my $file = $info->install_json`

    `install.json` file path

- `my $file = $info->mymeta_json`

    `MYMETA.json` file path

- `my $main_module = $info->main_module`

    main module name

- `my $version = $info->main_module_version`

    main module version

- `my $file = $info->main_module_file`

    main module file path

- `my $files = $info->files`

    file paths which is listed in `.packlist` file,
    note that paths are acutually `Cwd::abs_path()`-ed

- `my $hash = $info->install_json_hash`

    a hash reference for `install.json`

        my $info = Distribution::Metadata->new_from_module("LWP::UserAgent");
        my $install = $info->install_json_hash;
        $install->{version};  # 6.13
        $install->{dist};     # libwww-perl-6.13
        $install->{provides}; # a hash reference of providing modules
        ...

- `my $hash = $info->mymeta_json_hash`

    a hash reference for `MYMETA.json`

        my $info = Distribution::Metadata->new_from_module("LWP::UserAgent");
        my $meta = $info->mymeta_hash;
        $meta->{version};  # 6.13
        $meta->{abstract}; # The World-Wide Web library for Perl
        $meta->{prereqs};  # prereq hash
        ...

# SEE ALSO

[Module::Metadata](https://metacpan.org/pod/Module%3A%3AMetadata)

[App::cpanminus](https://metacpan.org/pod/App%3A%3Acpanminus)

# COPYRIGHT AND LICENSE

Copyright (C) 2015 Shoichi Kaji <skaji@cpan.org>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for Perl extension Distribution-Metadata

{{$NEXT}}

0.10  2023-05-07 00:36:47 JST
    - require perl v5.16

0.06  2017-11-24 22:11:07 JST
    - enable static install

0.05  2015-10-08 00:16:24 JST

    - fix bug agian; should have cache install.json per directories
    - document that paths are Cwd::abs_path()-ed (#6 reported by zachthompson)

0.04  2015-07-27 00:29:08 JST

    - fix bug; should have cache install.json per directories

0.03  2015-07-26 18:42:39 JST

    - introduce Factory class, which gives speed improvement
    - fix #2 (distributions which have no "main module" cannot be gathered)

0.02 2015-03-15T19:49:21Z

    - document that this module is still in development phase, and API will change without notice
    - speed improvements; cache core packlist, try to search naively guessed packlist and meta directory first
    - add name, version, distvname, author, pathname methods
    - other fixes (see #1)

0.01 2015-02-22T07:15:59Z

    - original version



Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.039.
use strict;
use warnings;

use 5.008001;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "gather distribution metadata in local",
  "AUTHOR" => "Shoichi Kaji",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Distribution-Metadata",
  "EXE_FILES" => [
    "script/which-meta"
  ],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.008001",
  "NAME" => "Distribution::Metadata",
  "PREREQ_PM" => {
    "CPAN::DistnameInfo" => 0,
    "CPAN::Meta" => 0,
    "ExtUtils::Packlist" => 0,
    "JSON" => 0,
    "Module::Metadata" => 0
  },
  "TEST_REQUIRES" => {
    "File::pushd" => 0,
    "Test::More" => "0.98"
  },
  "VERSION" => "0.04",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "CPAN::DistnameInfo" => 0,
  "CPAN::Meta" => 0,
  "ExtUtils::Packlist" => 0,
  "File::pushd" => 0,
  "JSON" => 0,
  "Module::Metadata" => 0,
  "Test::More" => "0.98"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


eg/remove-module.pl
#!/usr/bin/env perl
use strict;
use warnings;
use utf8;
use lib "../lib";
use Distribution::Metadata;
use Config;
use Getopt::Long qw(:config no_auto_abbrev no_ignore_case);
use File::Path 'rmtree';
use Pod::Usage 'pod2usage';
use IO::Handle;
require ExtUtils::MakeMaker;
sub prompt { ExtUtils::MakeMaker::prompt(@_) }
STDOUT->autoflush(1);

GetOptions
    "h|help" => sub { pod2usage(0) },
or pod2usage(1);

my $module = shift or pod2usage(1);

my $info = Distribution::Metadata->new_from_module($module);
die "Cannot find $module\n" unless $info->packlist;
if ($info->main_module eq "perl") {
    warn "ooooooooops, you are about to remove perl itself, stop!\n";
    exit;
}

my @unlink = grep defined, (
    @{ $info->files },
    $info->install_json, $info->mymeta_json, $info->meta_directory
);

warn "-> $_\n" for @unlink;
my $answer = prompt("=> Do you want to unlink the above files? (y/N)", "N");
exit if $answer !~ /^y$/i;

for my $entry (@unlink) {
    if (-f $entry) {
        if (unlink $entry) {
            warn "-> Removed $entry\n";
        } else {
            die "=> Failed to remove $entry: $!\n";
        }
    } elsif (-d $entry) {
        if (rmtree $entry) {
            warn "-> Removed $entry\n";
        } else {
            die "=> Failed to remove $entry";
        }
    }
}

__END__

=head1 SYNOPSIS

    > remove-module.pl MODULE

=cut


lib/Distribution/Metadata/Factory.pm
package Distribution::Metadata::Factory;
use strict;
use warnings;
use Distribution::Metadata;

sub new {
    my ($class, %option) = @_;
    my $inc = $option{inc} || \@INC;
    if ($option{fill_archlib}) {
        $inc = Distribution::Metadata->_fill_archlib($inc);
    }
    bless { inc => $inc, cache => {} }, $class;
}

sub create_from_module {
    my ($self, $module) = @_;
    local $Distribution::Metadata::CACHE = $self->{cache};
    Distribution::Metadata->new_from_module($module, inc => $self->{inc});
}

sub create_from_file {
    my ($self, $file) = @_;
    local $Distribution::Metadata::CACHE = $self->{cache};
    Distribution::Metadata->new_from_file($file, inc => $self->{inc});
}

1;

__END__

=for stopwords packlist

=encoding utf-8

=head1 NAME

Distribution::Metadata::Factory - create Distribution::Metadata objects with cache

=head1 SYNOPSIS

    use Distribution::Metadata::Factory;

    my $factory = Distribution::Metadata::Factory->new(inc => \@INC);

    my $info1 = $factory->create_from_module("Moose");
    my $info2 = $factory->create_from_module("Plack");
    my $info3 = $factory->create_from_file("/path/to/Moo.pm");

=head1 DESCRIPTION

This module creates Distribution::Metadata objects with cache.

If you creates many Distribution::Metadata objects,
then it may take quite a lot of time.
This is because C<< Distribution::Metadata->new >> scans C<@INC> directories,
parse packlist files, and parses install.json files many times.

Distribution::Metadata::Factory caches such results,
so you can create Distribution::Metadata objects even faster.

=head1 LICENSE

Copyright (C) 2015 Shoichi Kaji

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Shoichi Kaji E<lt>skaji@cpan.orgE<gt>

=cut



lib/Distribution/Metadata.pm
package Distribution::Metadata 0.06;
use v5.16;
use warnings;

use CPAN::DistnameInfo;
use CPAN::Meta;
use Config ();
use Cwd ();
use ExtUtils::Packlist;
use File::Basename ();
use File::Find ();
use File::Spec;
use JSON ();
use Module::Metadata;

use constant DEBUG => $ENV{PERL_DISTRIBUTION_METADATA_DEBUG};

my $SEP = qr{/|\\}; # path separater
my $ARCHNAME = $Config::Config{archname};

our $CACHE;

sub new_from_file {
    my ($class, $file, %option) = @_;
    $class->_new(%option, _module => {file => $file});
}

sub new_from_module {
    my ($class, $module, %option) = @_;
    $class->_new(%option, _module => {name => $module});
}

sub _new {
    my ($class, %option) = @_;
    my $module = $option{_module};
    my $inc = $option{inc} || \@INC;
    $inc = $class->_abs_path($inc);
    $inc = $class->_fill_archlib($inc) if $option{fill_archlib};
    my $metadata = $module->{file}
        ? Module::Metadata->new_from_file($module->{file}, inc => $inc)
        : Module::Metadata->new_from_module($module->{name}, inc => $inc);

    my $self = bless {}, $class;
    return $self unless $metadata;

    $module->{file} = $metadata->filename;
    $module->{name} = $metadata->name;
    $module->{version} = $metadata->version;

    my ($packlist, $files) = $class->_find_packlist($module->{file}, $inc);
    if ($packlist) {
        $self->{packlist} = $packlist;
        $self->{files}    = $files;
    } else {
        return $self;
    }

    my ($main_module, $lib) = $self->_guess_main_module($packlist);
    if ($main_module) {
        $self->{main_module} = $main_module;
        if ($main_module eq "perl") {
            $self->{main_module_version} = $^V;
            $self->{main_module_file} = $^X;
            $self->{dist} = "perl";
            my $version = "" . $^V;
            $version =~ s/v//;
            $self->{distvname} = "perl-$version";
            $self->{version} = $version;
            return $self;
        }
    } else {
        return $self;
    }

    my $archlib = File::Spec->catdir($lib, $ARCHNAME);
    my $main_metadata = Module::Metadata->new_from_module(
        $main_module, inc => [$archlib, $lib]
    );

    my ($find_module, $find_version);
    if ($main_metadata) {
        $self->{main_module_version} = $main_metadata->version;
        $self->{main_module_file} = $main_metadata->filename;
        $find_module = $main_metadata->name;
        $find_version = $main_metadata->version;
    } else {
        $find_module = $module->{name};
        $find_version = $module->{version};
    }

    my ($meta_directory, $install_json, $install_json_hash, $mymeta_json) = $class->_find_meta(
        $main_module, $find_module, $find_version,
        File::Spec->catdir($archlib, ".meta")
    );
    $self->{meta_directory}    = $meta_directory;
    $self->{install_json}      = $install_json;
    $self->{install_json_hash} = $install_json_hash;
    $self->{mymeta_json}       = $mymeta_json;
    $self;
}

sub _guess_main_module {
    my ($self, $packlist) = @_;
    my @piece = File::Spec->splitdir( File::Basename::dirname($packlist) );
    if ($piece[-1] eq $ARCHNAME) {
        return ("perl", undef);
    }

    my (@module, @lib);
    for my $i ( 1 .. ($#piece-2) ) {
        if ($piece[$i] eq $ARCHNAME && $piece[$i+1] eq "auto") {
            @module = @piece[ ($i+2) .. $#piece ];
            @lib    = @piece[ 0      .. ($i-1)  ];
            last;
        }
    }
    return unless @module;
    return ( _fix_module_name( join("::", @module) ), File::Spec->catdir(@lib) );
}

# ugly workaround for case insensitive filesystem
# eg: if you install 'Version::Next' module and later 'version' module,
# then version's packlist is located at Version/.packlist! (capital V!)
# Maybe there are a lot of others...
my @fix_module_name = qw(version Version::Next);
sub _fix_module_name {
    my $module_name = shift;
    if (my ($fix) = grep { $module_name =~ /^$_$/i } @fix_module_name) {
        $fix;
    } else {
        $module_name;
    }
}

sub _fill_archlib {
    my ($class, $incs) = @_;
    my %incs = map { $_ => 1 } @$incs;
    my @out;
    for my $inc (@$incs) {
        push @out, $inc;
        next if $inc =~ /$ARCHNAME$/o;
        my $archlib = File::Spec->catdir($inc, $ARCHNAME);
        if (-d $archlib && !$incs{$archlib}) {
            push @out, $archlib;
        }
    }
    \@out;
}

my $decode_install_json = sub {
    my $file = shift;
    my $content = do { open my $fh, "<", $file or next; local $/; <$fh> };
    JSON::decode_json($content);
};
sub _decode_install_json {
    my ($class, $file, $dir) = @_;
    if ($CACHE) {
        $CACHE->{install_json}{$dir}{$file} ||= $decode_install_json->($file);
    } else {
        $decode_install_json->($file);
    }
}

sub _find_meta {
    my ($class, $main_module, $module, $version, $dir) = @_;
    return unless -d $dir;

    my @install_json;
    if ($CACHE and $CACHE->{install_json_collected}{$dir}) {
        @install_json = keys %{$CACHE->{install_json}{$dir}};
    } else {
        @install_json = do {
            opendir my $dh, $dir or die "opendir $dir: $!";
            my @meta_dir = grep { !/^[.]{1,2}$/ } readdir $dh;
            grep -f, map { File::Spec->catfile($dir, $_, "install.json") } @meta_dir;
        };
        if ($CACHE) {
            $CACHE->{install_json}{$dir}{$_} ||= undef for @install_json;
            $CACHE->{install_json_collected}{$dir}++;
        }
    }

    # to speed up, first try distribution which just $module =~ s/::/-/gr;
    my $naive = do { my $dist = $main_module; $dist =~ s/::/-/g; $dist };
    @install_json = (
        (sort { $b cmp $a } grep {  /^$naive/ } @install_json),
        (sort { $b cmp $a } grep { !/^$naive/ } @install_json),
    );

    my ($meta_directory, $install_json, $install_json_hash, $mymeta_json);
    INSTALL_JSON_LOOP:
    for my $file (@install_json) {
        my $hash = $class->_decode_install_json($file, $dir);

        # name VS target ? When LWP, name is LWP, and target is LWP::UserAgent
        # So name is main_module!
        my $name = $hash->{name} || "";
        next if $name ne $main_module;
        my $provides = $hash->{provides} || +{};
        for my $provide (sort keys %$provides) {
            if ($provide eq $module
                && ($provides->{$provide}{version} || "") eq $version) {
                $meta_directory = File::Basename::dirname($file);
                $install_json = $file;
                $mymeta_json  = File::Spec->catfile($meta_directory, "MYMETA.json");
                $install_json_hash = $hash;
                last INSTALL_JSON_LOOP;
            }
        }
        DEBUG and warn "==> failed to find $module $version in $file\n";
    }

    return ($meta_directory, $install_json, $install_json_hash, $mymeta_json);
}

sub _naive_packlist {
    my ($class, $module_file, $inc) = @_;
    for my $i (@$inc) {
        if (my ($path) = $module_file =~ /$i $SEP (.+)\.pm /x) {
            my $archlib = $i =~ /$ARCHNAME$/o ? $i : File::Spec->catdir($i, $ARCHNAME);
            my $try = File::Spec->catfile( $archlib, "auto", $path, ".packlist" );
            return $try if -f $try;
        }
    }
    return;
}

# It happens that .packlist files are symlink path.
# eg: OSX,
# in .packlist: /var/folders/...
# but /var/folders/.. is a symlink to /private/var/folders
my $extract_files = sub {
    my $packlist = shift;
    [
        map  { Cwd::abs_path($_) } grep { -f }
        sort keys %{ ExtUtils::Packlist->new($packlist) || +{} }
    ];
};
sub _extract_files {
    my ($class, $packlist) = @_;
    if ($CACHE) {
        $CACHE->{packlist}{$packlist} ||= $extract_files->($packlist);
    } else {
        $extract_files->($packlist);
    }
}

sub _core_packlist {
    my ($self, $inc) = @_;
    for my $dir (grep -d, @$inc) {
        opendir my $dh, $dir or die "Cannot open dir $dir: $!\n";
        my ($packlist) = map { File::Spec->catfile($dir, $_) } grep {$_ eq ".packlist"} readdir $dh;
        return $packlist if $packlist;
    }
    return;
}

sub _find_packlist {
    my ($class, $module_file, $inc) = @_;

    if ($CACHE and my $core_packlist = $CACHE->{core_packlist}) {
        my $files = $class->_extract_files($core_packlist);
        if (grep {$module_file eq $_} @$files) {
            return ($core_packlist, $files);
        }
    }

    # to speed up, first try packlist which is naively guessed by $module_file
    if (my $naive_packlist = $class->_naive_packlist($module_file, $inc)) {
        my $files = $class->_extract_files($naive_packlist);
        if ( grep { $module_file eq $_ } @$files ) {
            DEBUG and warn "-> naively found packlist: $module_file\n";
            return ($naive_packlist, $files);
        }
    }

    my @packlists;
    if ($CACHE and $CACHE->{packlist_collected}) {
        @packlists = keys %{ $CACHE->{packlist} };
    } else {
        if (my $core_packlist = $class->_core_packlist($inc)) {
            push @packlists, $core_packlist;
            $CACHE->{core_packlist} = $core_packlist if $CACHE;
        }
        File::Find::find sub {
            return unless -f;
            return unless $_ eq ".packlist";
            push @packlists, $File::Find::name;
        }, grep -d, map { File::Spec->catdir($_, "auto") } @{$class->_fill_archlib($inc)};
        if ($CACHE) {
            $CACHE->{packlist}{$_} ||= undef for @packlists;
            $CACHE->{packlist_collected}++;
        }
    }

    for my $try (@packlists) {
        my $files = $class->_extract_files($try);
        if (grep { $module_file eq $_ } @$files) {
            return ($try, $files);
        }
    }
    return;
}

sub _abs_path {
    my ($class, $dirs) = @_;
    my @out;
    for my $dir (grep -d, @$dirs) {
        my $abs = Cwd::abs_path($dir);
        $abs =~ s/$SEP+$//;
        push @out, $abs if $abs;
    }
    \@out;
}

sub packlist            { shift->{packlist} }
sub meta_directory      { shift->{meta_directory} }
sub install_json        { shift->{install_json} }
sub mymeta_json         { shift->{mymeta_json} }
sub main_module         { shift->{main_module} }
sub main_module_version { shift->{main_module_version} }
sub main_module_file    { shift->{main_module_file} }
sub files               { shift->{files} }
sub install_json_hash   { shift->{install_json_hash} }

sub mymeta_json_hash {
    my $self = shift;
    return unless my $mymeta_json = $self->mymeta_json;
    $self->{mymeta_json_hash} ||= CPAN::Meta->load_file($mymeta_json)->as_struct;
}

sub _distnameinfo {
    my $self = shift;
    return unless my $hash = $self->install_json_hash;
    $self->{_distnameinfo} ||= CPAN::DistnameInfo->new( $hash->{pathname} );
}

for my $attr (qw(dist version cpanid distvname pathname)) {
    no strict 'refs';
    *$attr = sub {
        my $self = shift;
        return $self->{$attr} if exists $self->{$attr}; # for 'perl' distribution
        return unless $self->_distnameinfo;
        $self->_distnameinfo->$attr;
    };
}

# alias
sub name   { shift->dist }
sub author { shift->cpanid }

1;

__END__

=for stopwords .packlist inc pathname eg archname eq archlibs vname libwww-perl

=encoding utf-8

=head1 NAME

Distribution::Metadata - gather distribution metadata in local

=head1 SYNOPSIS

    use Distribution::Metadata;

    my $info = Distribution::Metadata->new_from_module("LWP::UserAgent");

    print $info->name;      # libwww-perl
    print $info->version;   # 6.13
    print $info->distvname; # libwww-perl-6.13
    print $info->author;    # ETHER
    print $info->pathname;  # E/ET/ETHER/libwww-perl-6.13.tar.gz

    print $info->main_module;         # LWP
    print $info->main_module_version; # 6.13
    print $info->main_module_file;    # path of LWP.pm

    print $info->packlist;       # path of .packlist
    print $info->meta_directory; # path of .meta directory
    print $info->install_json;   # path of install.json
    print $info->mymeta_json;    # path of MYMETA.json

    my $files = $info->files; # files which are listed in .packlist

    my $install_json_hash = $info->install_json_hash;
    my $mymeta_json_hash  = $info->mymeta_json_hash;

=head1 DESCRIPTION

(B<CAUTION>: This module is still in development phase. API will change without notice.)

Sometimes we want to know:
I<Where this module comes from? Which distribution does this module belong to?>

Since L<cpanm> 1.5000 (released 2011.10.13),
it installs not only modules but also their meta data.
So we can answer that questions!

Distribution::Metadata gathers distribution metadata in local.
That is, this module tries to gather

=over 4

=item * main module name, version, file

=item * C<.packlist> file

=item * C<.meta> directory

=item * C<install.json> file

=item * C<MYMETA.json> file

=back

Please note that as mentioned above, B<this module deeply depends on cpanm behavior>.
If you install cpan modules by hands or some cpan clients other than cpanm,
this module won't work.

=head1 HOW IT WORKS

Let me explain how C<< $class->new_from_module($module, inc => $inc) >> works.

=over 4

=item * Get C<$module_file> by

    Module::Metadata->new_from_module($module, inc => $inc)->filename.

=item * Find C<$packlist> in which C<$module_file> is listed.

=item * From C<$packlist> pathname (eg: ...auto/LWP/.packlist), determine C<$main_module> and main module search directory C<$lib>.

=item * Get C<$main_module_version> by

    Module::Metadata->new_from_module($main_module, inc => [$lib, "$lib/$Config{archname}"])->version

=item * Find install.json that has "name" eq C<$main_module>, and provides C<$main_module> with version C<$main_module_version>.

=item * Get .meta directory and MYMETA.json with install.json.

=back

=head2 CONSTRUCTORS

=over 4

=item C<< my $info = $class->new_from_module($module, inc => \@dirs, fill_archlib => $bool) >>

Create Distribution::Metadata instance from module name.

You can append C<inc> argument
to specify module/packlist/meta search paths. Default is C<\@INC>.

Also you can append C<fill_archlib> argument
so that archlibs are automatically added to C<inc> if missing.

Please note that, even if the module cannot be found,
C<new_from_module> returns a Distribution::Metadata instance.
However almost all methods returns false for such objects.
If you want to know whether the distribution was found or not, try:

    my $info = $class->new_from_module($module);

    if ($info->packlist) {
        # found
    } else {
        # not found
    }

=item C<< my $info = $class->new_from_file($file, inc => \@dirs, fill_archlib => $bool) >>

Create Distribution::Metadata instance from file path.
You can append C<inc> and C<fill_archlib> arguments too.

Also C<new_from_file> retunes a Distribution::Metadata instance,
even if file cannot be found.

=back

=head2 METHODS

Please note that the following methods return false
when appropriate modules or files cannot be found.

=over 4

=item C<< my $name = $info->name (alias: $info->dist) >>

distribution name (eg: C<libwww-perl>)

=item C<< my $version = $info->version >>

distribution version (eg: C<6.13>)

=item C<< my $distvname = $info->distvname >>

distribution vname (eg: C<libwww-perl-6.13>)

=item C<< my $author = $info->author (alias: $info->cpanid) >>

distribution author (eg: C<ETHER>)

=item C<< my $pathname = $info->pathname >>

distribution pathname (eg: C<E/ET/ETHER/libwww-perl-6.13.tar.gz>)

=item C<< my $file = $info->packlist >>

C<.packlist> file path

=item C<< my $dir = $info->meta_directory >>

C<.meta> directory path

=item C<< my $file = $info->install_json >>

C<install.json> file path

=item C<< my $file = $info->mymeta_json >>

C<MYMETA.json> file path

=item C<< my $main_module = $info->main_module >>

main module name

=item C<< my $version = $info->main_module_version >>

main module version

=item C<< my $file = $info->main_module_file >>

main module file path

=item C<< my $files = $info->files >>

file paths which is listed in C<.packlist> file,
note that paths are acutually C<< Cwd::abs_path() >>-ed

=item C<< my $hash = $info->install_json_hash >>

a hash reference for C<install.json>

    my $info = Distribution::Metadata->new_from_module("LWP::UserAgent");
    my $install = $info->install_json_hash;
    $install->{version};  # 6.13
    $install->{dist};     # libwww-perl-6.13
    $install->{provides}; # a hash reference of providing modules
    ...

=item C<< my $hash = $info->mymeta_json_hash >>

a hash reference for C<MYMETA.json>

    my $info = Distribution::Metadata->new_from_module("LWP::UserAgent");
    my $meta = $info->mymeta_hash;
    $meta->{version};  # 6.13
    $meta->{abstract}; # The World-Wide Web library for Perl
    $meta->{prereqs};  # prereq hash
    ...

=back

=head1 SEE ALSO

L<Module::Metadata>

L<App::cpanminus>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2015 Shoichi Kaji E<lt>skaji@cpan.orgE<gt>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut
